packages/flutter/lib/src/widgets/layout_builder.dart

```diff
@@ -0,0 +1,205 @@
+
+  /// Called at layout time to construct the widget tree. The builder may not
+  /// return null.
+  final LayoutWidgetBuilder builder;
+
+  @override
+  _LayoutBuilderElement createElement() => new _LayoutBuilderElement(this);
+
+  @override
+  _RenderLayoutBuilder createRenderObject(BuildContext context) => new _RenderLayoutBuilder();
+}
+
+class _RenderLayoutBuilder extends RenderBox with RenderObjectWithChildMixin<RenderBox> {
+  _RenderLayoutBuilder({ LayoutCallback callback }) : _callback = callback;
+
+  LayoutCallback get callback => _callback;
+  LayoutCallback _callback;
+  void set callback(LayoutCallback value) {
+    if (value == _callback)
+      return;
+    _callback = value;
+    markNeedsLayout();
+  }
+
+  double getIntrinsicWidth(BoxConstraints constraints) => constraints.constrainWidth();
+
+  double getIntrinsicHeight(BoxConstraints constraints) => constraints.constrainHeight();
+
+  @override
+  double getMinIntrinsicWidth(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsValid());
+    return getIntrinsicWidth(constraints);
+  }
+
+  @override
+  double getMaxIntrinsicWidth(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsValid());
+    return getIntrinsicWidth(constraints);
+  }
+
+  @override
+  double getMinIntrinsicHeight(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsValid());
+    return getIntrinsicHeight(constraints);
+  }
+
+  @override
+  double getMaxIntrinsicHeight(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsValid());
+    return getIntrinsicHeight(constraints);
+  }
+
+  @override
+  bool get sizedByParent => true;
+
+  @override
+  void performResize() {
+    size = constraints.biggest;
+  }
+
+  @override
+  void performLayout() {
+    if (callback != null)
+      invokeLayoutCallback(callback);
+  }
+
+  @override
+  bool hitTestChildren(HitTestResult result, { Point position }) {
+    return child?.hitTest(result, position: position) ?? false;
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    if (child != null)
+      context.paintChild(child, offset);
+  }
+}
+
+class _LayoutBuilderElement extends RenderObjectElement {
+  _LayoutBuilderElement(LayoutBuilder widget) : super(widget);
+
+  @override
+  LayoutBuilder get widget => super.widget;
+
+  @override
+  _RenderLayoutBuilder get renderObject => super.renderObject;
+
+  Element _child;
+
+  @override
+  void visitChildren(ElementVisitor visitor) {
+    if (_child != null)
+      visitor(_child);
+  }
+
+  @override
+  void mount(Element parent, dynamic newSlot) {
+    super.mount(parent, newSlot);    // Creates the renderObject.
+    renderObject.callback = _layout; // The _child will be built during layout.
+  }
+
+  @override
+  void update(LayoutBuilder newWidget) {
+    super.update(newWidget);
+    assert(widget == newWidget);
+    renderObject.callback = _layout;
+    renderObject.markNeedsLayout();
```

Comment:
I thought I needed to unconditionally call for another layout because the LayoutBuilder's builder function, even if it hasn't changed, might return a different widget?